### DISCUSSION TITLE
Typesetting per Tesi

### DISCUSSION THREAD

**Discussion by cristian-5** [csunibo/csunibo.github.io#12] [Domande] [answered]
Discutevo con @foxyseta di un progetto che riguarda la realizzazione di un tool in grado di soppiantare **TeX** e le sue varianti che ormai hanno bisogno di una boccata d'aria.

### Impaginare è un problema?
Questo lo sappiamo perché ci sono studenti che addirittura arrivano a pagare per servizi di impaginazione che impostano tutto secondo le linee guida dell'Università.
Lo strumento più usato è **Microsoft Word** (prendendo in considerazione tutte le facoltà).
Ci sono molte cose tecniche che sono difficili da aggiustare (ad esempio i margini e i paragrafi). Basti pensare che spesso per fare i paragrafi c'è chi manda a capo e lascia una riga vuota, per poi modificare la grandezza del testo di quella riga e ridurla allo spazio voluto.

### STEM e TeX, LaTeX
Se non conosci **TeX**:

> TeX è un software che permette di imaginare tesi e pubblicazioni che viene usato maggiormente in ambito STEM in quanto ti permette di generare diagrammi di ogni tipo e fare la stampa di espressioni matematiche come quelle dei libri (per capirci).

Sicuramente hai avuto, in qualche punto della tua carriera, un file generato con questo software, la cui caratteristica principale è il font estremamente leggibile.

### Perché TeX è arrivato al capolinea?
La sintassi di TeX è assurdamente complessa e non perdona. Un'errore e non funziona più nulla. Capire qual'è l'errore è difficilissimo in quanto sputa centinaia di righe di errore effetto domino e davvero non si capisce nulla.
(Se stai pensando che non è così male perché tu riesci a individuare e capire "OVERFULL HBOX" tra 1000 righe di codice allora le cose cambiano se fai scrivere la tesi ad un povero studente di qualche altra facoltà poco informatizzata.)

# Cosa fare?
Adottare tutte le funzionalità principali di TeX, eliminare la sintassi ortodossa e introdurre una sintassi basata su *MarkDown* che bene o male è decisamente più semplice da imparare. Il markdown viene poi trasformato in `html` per stampa e poi salvato in **pdf** da qualche middleware di traduzione.

# Come rende le cose più semplici?
Per impaginare basterebbe specificare un foglio di stile `css` che ogni università può fornirti.
Il markdown ha il supporto per le espressioni matematiche, i link, le immagini, gli elenchi, le tabelle, e persino il syntax highlighting del codice.\
Sarebbe molto molto più veloce di TeX che è un linguaggio che sfrutta macro su macro e 80 pagine le compila in 13 secondi.\
Non ci sarebbero più errori non ovvi. Se fai un'errore, il `pdf` sarà sbagliato (non corrotto attenzione) all'esatta sezione dove c'è qualcosa che non va. Per lo più `html` è in grado di sanare tanti ma tanti errori senza gravi modifiche al risultato finale.\
Per non parlare del fatto che il software sarebbe in grado di creare indice, appendice e note a fine pagina in modo automatico.

# Prototipo
Al momento si cercano volontari, ho iniziato il prototipo in *javascript* per **Deno** che già promette bene e da qualche risultato.

**Comment by foxyseta** [comment]
Noi al momento per tradurre da MD a PDF e HTMl usiamo `pandoc`, che sicuramente può essere migliorato. Molto probabilmente però alternative a `pandoc` basate su sintassi MD che soddisfano i tuoi requisiti esistono già. Penso sia urgente prima fare una ricerca e poi se nessuna alternativa soddisfa considerare la più vicina a quelle che ci/ti possono interessare e partire da una sua fork. Purtroppo **non** vediamo di buon occhio i progetti in Node.

**Reply by cristian-5** [reply]
Vedi qui il problema è che pandoc non è in grado di impaginare testi che poi vanno stampati... come faresti ad esempio l'interruzione di pagina? per non parlare del fatto che dovresti usare altri software per fare grafici e tabelle, e note a pie di pagina. Qui l'intento è soppiantare latex ma mantenere lo stesso output con una sintassi semplificata. Poi non ho detto di usare node. Io pensavo di usare Deno che è l'evoluzione di node che mette al centro la sicurezza e l'usabilità. Addio anche alla `node_modules` folder, non più necessaria.
Il javascript è ideale per il markdown perché se uno vuole aggiungere un'estensione per fare un componente suo, può farlo eseguire sia lato preprocessore sia lato client-rendering senza dover imparare un altro linguaggio. Qui di certo il problema non è il linguaggio (che possiamo sempre cambiare) ma la creazione di un tool che possano usare tutti.

**Reply by cristian-5** [reply]
deduco che non ti piacciono i linguaggi weakly typed, sono pienamente d'accordo. io faccio linguaggi e compilatori per hobby e riesco ad apprezzare meglio un linguaggio con compile-time type-checking tuttavia questo è un prototipo al momento e la priorità è mettere insieme un qualcosa che funzioni, anche per grazia divina.

**Reply by foxyseta** [reply]
Chiaro tutto. Hai già cercato se esitano progetti simili?

**Reply by cristian-5** [reply]
Markdown-It da cui si può attingere e Puppetteer per fare il pdf altrimenti ci sarebbe Markdown to PDF ma non possiamo modificarlo facilmente quindi ho iniziato con i primi due.

**Comment by VaiTon** [comment, marked as answer]
https://asciidoc.org/

**Reply by cristian-5** [reply]
Questo mi sembra molto valido, e anche estremamente veloce nel compilare. Un po' complicato modificare lo stile del pdf dato che le proprietà vanno specificate in un file yaml. Magari facciamo così @foxyseta , sviluppiamo una repo con le istruzioni per usare questo software e prepariamo noi i file di stile con le specifiche dell'università. In questo modo gli studenti possono scaricare i file dello starter pack e scrivere le tesi con questo senza doversi preoccupare dell'impaginazione.

**Reply by cristian-5** [reply]
ma soprattutto, niente javascript

**Reply by foxyseta** [reply]
Mi sembrano tutte buone idee.

**Reply by foxyseta** [reply]
Faccio una nuova repo template su csunibo e ti invito?

**Reply by cristian-5** [reply]
perfetto

**Reply by foxyseta** [reply]
Invitato come membro del gruppo degli esperti di composizione di documenti (il che ti dà accesso come manutentore a https://github.com/csunibo/asciidoc-thesis nonché ad altre repo tematicamente affini).
